### PR TITLE
Fix recent regressions with app hosts, test specs, & modular headers

### PR DESCRIPTION
**Pull request by segiddins**
- [x] Tests for all three issues being fixed

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* [WIP] Fix recent regressions with app hosts, test specs, & modular headers  
  [segiddins](https://github.com/segiddins)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/cocoapods/target/pod_target.rb [hunk 1]
```diff
@@ -209,7 +209,15 @@ def static_framework?
     def defines_module?
       return @defines_module if defined?(@defines_module)
       return @defines_module = true if uses_swift? || requires_frameworks?
-      return @defines_module = true if target_definitions.all? { |td| td.build_pod_as_module?(pod_name) }
+
+      explicit_target_definitions = target_definitions.select { |td| td.dependencies.any? { |d| d.root_name == pod_name } }
+      tds_by_answer = explicit_target_definitions.group_by { |td| td.build_pod_as_module?(pod_name) }
+
+      if tds_by_answer.size > 1
+        UI.warn "Ambiguous whether to build #{self} as a module.\n\t- #{tds_by_answer.map { |a, t| "#{t.to_sentence} say #{a ? 'to' : 'not to'}" }.join("\n\t- ")}\nNot building as a module."
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: this message is not super user friendly. Also wrap `#{self}` with backticks as we do everywhere else?

File: lib/cocoapods/target/pod_target.rb [hunk 2]
```diff
@@ -419,15 +427,18 @@ def _add_recursive_dependent_targets(set)
...
+      @recursive_test_dependent_targets ||= {}
+      @recursive_test_dependent_targets[test_spec] ||= _add_recursive_test_dependent_targets(test_spec, Set.new).to_a
     end
 
-    def _add_recursive_test_dependent_targets(set)
-      test_dependent_targets_by_spec_name.each_value do |dependent_targets|
-        dependent_targets.each do |target|
-          target._add_recursive_dependent_targets(set) if set.add?(target)
-        end
+    def _add_recursive_test_dependent_targets(test_spec, set)
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: is this private? can we move it under `private` if not?

**Comment by dnkoutso** [comment]
Taking over this branch and will land it so we can ship 1.6.0.beta.1.

File: spec/unit/installer/xcode/pods_project_generator/pod_target_installer_spec.rb [hunk 3]
```diff
@@ -163,13 +163,13 @@ class PodsProjectGenerator
                 @project = Project.new(config.sandbox.project_path)
                 config.sandbox.project = @project
 
-                @coconut_spec = fixture_spec('coconut-lib/CoconutLib.podspec')
+                @watermelon_spec = fixture_spec('watermelon-lib/WatermelonLib.podspec')
```
Comment (dnkoutso) [inline spec/unit/installer/xcode/pods_project_generator/pod_target_installer_spec.rb]: re-wrote this to use `WatermelonLib` fixture which has two test specs and helps us test test spec split.

**Comment by dnkoutso** [comment]
Integration specs https://github.com/CocoaPods/cocoapods-integration-specs/pull/179

File: lib/cocoapods/target/pod_target.rb (hunk 2 above)
Comment (segiddins) [reply lib/cocoapods/target/pod_target.rb]: its `protected`

File: lib/cocoapods/target/pod_target.rb (hunk 1 above)
Comment (segiddins) [reply lib/cocoapods/target/pod_target.rb]: sure!

File: lib/cocoapods/target/pod_target.rb [hunk 4]
```diff
@@ -419,15 +427,18 @@ def _add_recursive_dependent_targets(set)
...
+      raise ArgumentError, 'Must give a test spec' unless test_spec
+      # raise ArgumentError, "Attempting to add recursive test dependent targets for #{self} but #{test_spec} is not a known test spec!" unless dependent_targets = test_dependent_targets_by_spec_name[test_spec.name]
```
Comment (segiddins) [inline lib/cocoapods/target/pod_target.rb]: commented-out code. ideally we could use it, but it broke a bunch of unit tests

File: lib/cocoapods/target/pod_target.rb [hunk 5]
```diff
@@ -214,7 +214,9 @@ def defines_module?
       tds_by_answer = explicit_target_definitions.group_by { |td| td.build_pod_as_module?(pod_name) }
 
       if tds_by_answer.size > 1
-        UI.warn "Ambiguous whether to build #{self} as a module.\n\t- #{tds_by_answer.map { |a, t| "#{t.to_sentence} say #{a ? 'to' : 'not to'}" }.join("\n\t- ")}\nNot building as a module."
+        UI.warn "Unable to determine whether to build `#{label}` as a module due to a conflict " \
+          "between the following target definitions: #{tds_by_answer.map { |_, td| "`#{td.to_sentence}`" }.join(', ')}. " \
```
[the remark this answers is not in the document]
Comment (dnkoutso) [reply lib/cocoapods/target/pod_target.rb]: ok will update

**Comment by dnkoutso** [comment]
green! Will update warning message now

**Comment by dnkoutso** [comment]
Updating message then merging on green.

**Comment by dnkoutso** [comment]
Ready.

**Pull request merged**